src/jvm/main/org/jetbrains/kotlinx/lincheck/util/AnalysisSections.kt

```diff
@@ -11,8 +11,91 @@
+ * These sections provide various analysis guarantees and characteristics.
+
+ * The sections are ordered by the strength of the provided guarantees, from weakest to strongest:
+ *   [NORMAL], [SILENT], [SILENT_PROPAGATING], [ATOMIC], [IGNORE].
+ * Guarantees of each particular section are documented separately.
+ *
+ * Section types can be split into two categories: local and propagating (see [isCallStackPropagating]).
+ *
+ *   - Local section is only applied to the current method in the call stack.
+ *     Other methods down the call stack can reside in different analysis section types,
+ *     including weaker section types.
+ *     Thus, it is possible to have an alternating sequence of different section types in the call stack.
+ *
+ *   - Propagating sections propagate down the call stack.
+ *     Other methods down the call stack can only reside in the same analysis section, or a stronger one.
+ *
+ * Local sections are required to support methods like `ConcurrentHashMap.computeIfAbsent(key, lambda)`.
+ * The method `computeIfAbsent` itself can be trusted and
+ * not analyzed by the framework in the user code by default.
+ * However, the user-provided lambda is not trusted and needs to be analyzed.
+ *
+ * Thus, `computeIfAbsent` can be put into a (local, non-propagating) silent section.
+ * As such, code inside `computeIfAbsent` will be muted,
+ * but if this method calls the provided lambda, it still will be analyzed fully.
+ *
+ * Local sections:
+ *   - [NORMAL]
+ *   - [SILENT]
+ * Propagating sections:
+ *   - [SILENT_PROPAGATING]
+ *   - [ATOMIC]
+ *   - [IGNORE]
+ */
+internal enum class AnalysisSectionType {
+
+    /**
+     * Normal section without special handling.
+     * Inside normal sections, all events are tracked.
+     * All the events occurring inside a normal section are added to the trace,
+     * unless other factors prevent it.
+     * Thread switch points can occur arbitrarily within a normal section.
+     */
+    NORMAL,
+
+    /**
+     * Silent sections are used to mute analysis.
+     * Note that events are still tracked inside silent sections, but
+     * they are not added to the trace by default.
+     * Thread switch points can occur inside a silent section only if they are forced.
+     * For instance, because of some blocking synchronization primitive,
+     * like an attempt to acquire a monitor that is already held by another thread.
+     */
+    SILENT,
+
+    /**
+     * Same as the silent section, but propagates down the call stack.
+     */
+    SILENT_PROPAGATING,
+
+    /**
+     * Code inside an atomic section behaves like an atomic indivisible operation.
+     * Technically, the atomic section is a stronger version of the nested silent section,
+     * with an additional guarantee that thread switch points cannot occur inside an atomic section at all.
+     * TODO: "no-switch-points inside atomic section" guarantee is not checked currently.
+     */
+    ATOMIC,
```

Comment:
Is there any difference between atomic and silent? Do we need both?

Reply:
The only (and maybe important) difference I see is that for atomic sections, it's fine not to transform the corresponding byte-code -- we just switch off the analysis

Reply:
> it's fine not to transform the corresponding byte-code -- we just switch off the analysis

I think we still might want to track some events inside atomic sections --- that is potentially blocking events (like `lock`, `park`, etc). This would be necessary to check that these events do not force blocking and thus a thread switch inside an atomic section (to verify that this section is indeed atomic and does not have any thread switches at all).